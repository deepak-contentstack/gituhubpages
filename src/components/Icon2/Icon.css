@import '../../../tokens/variables.scss';
@import '../../../tokens/tokens.css';
@import '../../static/styles/css/typography.css';
@import '../../static/styles/css/variable.css';
@import '../../static/styles/css/common.css';

.IconStoriesWrapper {
    align-items: center;
    display: flex;
    flex-wrap: wrap;
    font-family: $font-family-primary;
}

.Icon--original {
    height: auto;
    width: auto;
}

.Icon--extraSmall {
    height: $space-10;
    width: $space-10;
}

.Icon--mini {
    height: $space-12;
    width: $space-12;
}

.Icon--tiny {
    height: $space-14;
    width: $space-14;
}

.Icon--small {
    height: $space-18;
    width: $space-18;
}

.Icon--medium {
    height: $space-24;
    width: $space-24;
}

.Icon--large {
    height: $space-36;
    width: $space-36;
}

.Icon--secondary {
    path {
        stroke: $color-base-orange-5;
    }
}

.Icon--white {
    path {
        stroke: $color-base-white-5;
    }
}

/* CSS for optional div container around the icon */

.Icon__shadow {
    background: $color-brand-secondary-lighter;
    border-radius: $radii-50;
    cursor: pointer;
    
    &-mini {
        height: $space-24;
        width: $space-24;
    }
    
    &-tiny {
        height: $space-30;
        width: $space-30;
    }

    &-medium {
        height: $space-40;
        width: $space-40;
    }
}

/* Default, Hover, Active icon variations (test) */

.Icon--hover--primary {
    path, circle, rect {
        stroke: $color-base-purple-5;
    }
}

.Icon--hover--secondary {
    path[stroke], circle[stroke], rect[stroke] {
        stroke: $color-base-gray-20;
    }
    path[fill], circle[fill], rect[fill] {
        fill: $color-base-gray-20;
    }
}

.Icon--active {
    fill: $color-base-purple-5;
    path {
        fill: $color-base-purple-5;
        stroke: $color-base-purple-5;
    }
}

/* css for custom hover and active state in storybook stories*/
.custom-hover-red {
    path {
        stroke: $color-border-warning;
    }
}

.custom-active-red {
    fill: $color-border-warning;
    path {
        fill: $color-border-warning;
        stroke: $color-border-warning;
    }
}

/* ---Icon wrapper CSS */
.icon-spacing {
    background: #eaeaea;
    border-radius: 3px;
    font-size: 14px;
    margin: 5px;
    padding: $space-5;
    padding-left: 13px;
    width: 310px;
    svg{
        vertical-align: middle;
    }
    .icon-flex{
        align-items: center;
        display: flex;
        .icon-name{
            cursor: pointer;
            padding: $space-5;
            padding-left: 13px;
        }
    }
}