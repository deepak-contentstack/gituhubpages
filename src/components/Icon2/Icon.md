<!-- # Icon

Icons are often used when explaining a process, as you can visually represent each step.

## When to use

When you want to help user visually understand each step of a process and facilitate the visual separation and understanding of the elements they contain.

- **Default** - This is the default state of the icon.
  e.g. &lt;Icon icon='Star' &gt;

- **Hover** - To be used for hover state of icon by setting hover property to true.
  e.g. &lt;Icon icon='Star' hover='true' &gt;

- **HoverType** - To be used to set the hover behaviour of the icon. Primary by default.

- **container** - To be used to wrap the SVG in a div element of specific size. Necessary
when hover type is secondary.

- **Active** - To be used for active state of icon by setting active property to true.
  e.g. &lt;Icon icon='Star' active='true' &gt;

- **activeIcon** - An alternative to the above active. Specify a different Icon that replaces the current Icon when the Icon is in active mode. **This still needs the active prop to be passed.**
  e.g. &lt;Icon icon='PersonalizationExperiences' active={activeConditionhere} activeIcon='PersonalizationExperiencesHover' &gt;

### Additional properties

- **fill** - Can be used to set custom fill color to corresponding svg element.
  e.g. &lt;Icon icon='Star' fill='pink' &gt;

- **stroke** - Can be used to set custom stroke color to corresponding svg element.
  e.g. &lt;Icon icon='Star' stroke='pink' &gt;

- **width** - Can be used to set custom width to corresponding svg element.
  e.g. &lt;Icon icon='Star' width='100' &gt;

- **height** - Can be used to set custom height to corresponding svg element.
  e.g. &lt;Icon icon='Star' height='100' &gt;

- **data** - Can be used to render a string or an Element besides Icon.
  e.g. &lt;Icon icon='Star' data="data as string" &gt;
  e.g. &lt;Icon icon='Star' data=<>data as markup</> &gt;

### How to add new Icons to the Icon component

Place the new svg file in the svg folder. Add a property(newly added svg filename) to the object('iconName') in constants.ts file. Also import the new svg file and add the imported svg in object('iconComponents') in constants.ts file.

**Naming convention** - Imported svg file should follow CamelCasing, no usage of underscore in between name.

#### e.g import **IconName** from './svg/IconName.svg -->
