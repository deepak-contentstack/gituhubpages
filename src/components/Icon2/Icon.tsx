import React, { CSSProperties, useState, useEffect } from 'react'
import cn from 'classnames'

import { iconName, iconComponents } from './constants'

import './Icon.css'
import Tooltip from '../Tooltip/Tooltip'

export type IconType = keyof typeof iconName
// export type IconColorType = 'primary' | 'secondary' | 'white'

export type IconSize = 'original' | 'tiny' | 'mini' | 'small' | 'large' | 'medium' | 'extraSmall'

// Default prop value was not working with this way added default props with another way

// const defaultProps = {
//   size: 'original',
//   withTooltip: false,
//   testId: 'cs-Icon'
// }
// type DefaultProps = Readonly<typeof defaultProps>

export type IconProps = {
  icon: string
  /**
   * sizes can be 'original' | 'tiny' | 'small' | 'large' | 'medium'
   */
  size?: IconSize
  active?: Boolean
  hover?: Boolean
  hoverType?: 'primary' | 'secondary'
  shadow?: 'mini' | 'tiny' | 'medium'
  /**
   * custom css class for active state of icon
   */
  activeClassName?: string
  /**
   * custom css class for hover state of icon
   */
  hoverClassName?: string
  className?: string
  style?: CSSProperties
  fill?: string
  stroke?: string
  height?: Number | string
  width?: Number | string
  id?: string
  onClick?: Function
  disabled?: boolean
  activeIcon?: string
  viewBox?: string
  data?: string | React.ReactNode
  testId?: string
  // withTooltip?: boolean
  // tooltipContent?: string | React.ReactNode
  // tooltipPosition?: string
  withTooltip?: boolean
  tooltipContent?: string | React.ReactNode
  tooltipPosition?: string
}
// & Partial<DefaultProps>

//Should be replaced with Fallback Icon below
const FallbackIcon = () => null

export const Icon = (props: IconProps) => {
  const {
    icon,
    size,
    hover,
    hoverType = "primary",
    shadow,
    active,
    activeClassName,
    hoverClassName,
    id,
    className,
    style,
    activeIcon,
    data,
    testId,
    // withTooltip,
    // tooltipContent,
    // tooltipPosition,
    withTooltip,
    tooltipContent,
    tooltipPosition,
    ...otherProps
  } = props

  /**
   * To handle test env,
   * while testing, iconComponents[icon] = {}
   * So this must return Something valid component
   **/
  const [iconToDisplay, setIconToDisplay] = useState<string>(icon)

  const Element =
    typeof iconComponents[iconToDisplay] === 'object' && !Object.keys(iconComponents[iconToDisplay]).length
      ? FallbackIcon
      : iconComponents[iconToDisplay] || FallbackIcon

  /**
   * ref:-
   * State "iconColor" has three values i.e. 1)default, 2)active, and 3)hover.
   * 1)'default': original svg(default state of icon),
   * 2)'active': fill and stroke property applied to original svg(active state of icon)
   * 3)'hover': only stroke property applied to original svg(hover state of icon).
   **/
  const [iconState, setIconState] = useState<"default" | "hover" | "active">(active && !activeIcon ? 'active' : 'default')

  useEffect(() => {
    if (active) {
      if (activeIcon) {
        setIconToDisplay(activeIcon)
        if (iconState === 'hover') {
          setIconState('default')
        }
      }
      else {
        setIconState('active')
      }
    }
    else {
      setIconToDisplay(icon)
      setIconState('default')
    }
    // (used to disable es lint warning for next line)
    // eslint-disable-next-line react-hooks/exhaustive-deps 
  }, [active, icon])

  const classNames = cn(className, {
    [`Icon--${size}`]: size,
    [`Icon--hover--${hoverType}`]: iconState === 'hover',
    [`Icon--active`]: iconState === 'active',
    [`${activeClassName}`]:
      iconState === 'active' ? (typeof activeClassName === 'string' ? true : false) : false,
    [`${hoverClassName}`]:
      iconState === 'hover' ? (typeof hoverClassName === 'string' ? true : false) : false,
  })

  const shadowClassnames = cn("flex-center", "flex-v-center", {
    [`Icon__shadow`]: iconState === 'hover',
    [`Icon__shadow-${shadow}`]: shadow
  })

  const mouseEnterHandle = (event: any) => {
    if (!active) {
      if (hover) {
        setIconState('hover')
      }
    }
  }
  const mouseLeaveHandle = (event: any) => {
    if (!active) {
      if (hover) {
        setIconState('default')
      }
    }
  }

  const iconElement = <Element
    onMouseEnter={mouseEnterHandle}
    onMouseLeave={mouseLeaveHandle}
    className={classNames}
    style={style}
    name={icon}
    id={id}
    {...otherProps}
  />


  if (data) {
    return shadow ? (
      <div className="flex-v-center" data-test-id={testId}>
        <div className={shadowClassnames}>{iconElement}</div>
        {data && data}
      </div>
    ) : (
      <div className="flex-v-center" data-test-id={testId}>{iconElement}{data && data}</div>
    )
  } else {
    return shadow ? (
      <div
        className={shadowClassnames}
        data-test-id={testId}
      >
        {withTooltip ? <Tooltip content={tooltipContent ? tooltipContent : ''} position={tooltipPosition ? tooltipPosition : 'top'}>{iconElement}</Tooltip> :
          iconElement
        }
      </div>
    ) : (
      withTooltip ? <Tooltip content={tooltipContent ? tooltipContent : ''} position={tooltipPosition ? tooltipPosition : 'top'}>{iconElement}</Tooltip> :
        iconElement

    )
  }
}

Icon.defaultProps = {
  size: 'original',
  withTooltip: false,
  testId: 'cs-Icon'
}
export default Icon