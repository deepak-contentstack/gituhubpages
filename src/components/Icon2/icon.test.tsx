import React from 'react'
import renderer from 'react-test-renderer'

import Icon from './Icon'

const iconName = {
  Assets: 'Assets',
  Content_Models: 'Content_Models',
  Dashboard: 'Dashboard',
  Entries: 'Entries',
  Releases: 'Releases',
  Tasks: 'Tasks',
}

describe('Icon', () => {
  it('render default icons', () => {
    const DefaultIcons = () => (
      <div
        style={{
          columns: '4',
          width: '800px',
        }}
      >
        {Object.keys(iconName).map((icon: any, idx: number) => {
          return (
            <div className="icon-spacing" key={idx}>
              <Icon icon={icon} size="large" />
            </div>
          )
        })}
      </div>
    )
    const tree = renderer.create(<DefaultIcons />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('render Icons with different icon state', () => {
    const IconState = () => (
      <>
        <div>ICONS DEFAULT :</div>
        <div
          style={{
            columns: '4',
            width: '800px',
          }}
        >
          {Object.keys(iconName).map((icon: any, idx: number) => {
            return (
              <div className="icon-spacing" key={idx}>
                <Icon icon={icon} size="large" />
              </div>
            )
          })}
        </div>
        <div>ICONS HOVER :</div>
        <div
          style={{
            columns: '4',
            width: '800px',
          }}
        >
          {Object.keys(iconName).map((icon: any, idx: number) => {
            return (
              <div className="icon-spacing" key={idx}>
                <Icon icon={icon} hover={true} size="large" />
              </div>
            )
          })}
        </div>
        <div>ICONS PRESSED/SELECTED :</div>
        <div
          style={{
            columns: '4',
            width: '800px',
          }}
        >
          {Object.keys(iconName).map((icon: any, idx: number) => {
            return (
              <div className="icon-spacing" key={idx}>
                <Icon icon={icon} active={true} size="large" />
              </div>
            )
          })}
        </div>
      </>
    )
    const tree = renderer.create(<IconState />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('render icons with different sizes', () => {
    const IconSizes = () => (
      <div>
        <div>
          <Icon icon="Releases" size="tiny" /> Tiny
        </div>
        <div>
          <Icon icon="Releases" size="small" /> Small
        </div>
        <div>
          <Icon icon="Releases" size="large" /> Large
        </div>
      </div>
    )
    const tree = renderer.create(<IconSizes />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('render icon with fill property', () => {
    const IconWithFillProperty = () => (
      <div>
        <div>
          <Icon icon="Releases" size="large" />
          Default Icon
        </div>
        <div>
          <Icon icon="Releases" size="large" fill="red" />
          Icon with fill property
        </div>
      </div>
    )
    const tree = renderer.create(<IconWithFillProperty />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('render icon with stroke property', () => {
    const IconWithStrokeProperty = () => (
      <div>
        <div>
          <Icon icon="Dropdown" size="large" />
          Default Icon
        </div>
        <div>
          <Icon icon="Dropdown" size="large" stroke="red" />
          Icon with stroke property
        </div>
      </div>
    )
    const tree = renderer.create(<IconWithStrokeProperty />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('render icon with height and width property', () => {
    const IconWithHeightAndWidthProperty = () => (
      <div>
        <div>
          <Icon icon="Dropdown" size="large" />
          Default Icon
        </div>
        <div>
          <Icon icon="Dropdown" height="100" width="100" />
          Icon with height and width property
        </div>
      </div>
    )
    const tree = renderer.create(<IconWithHeightAndWidthProperty />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('render Icons with data', () => {
    const dataAsText = 'data as string'
    const DataAsMarkup = <>data as markup</>
    const IconWithData = () => (
      <>
        <div className="flex IconStoriesWrapper">
          <Icon icon="Releases" data={dataAsText} />
        </div>
        <div className="flex IconStoriesWrapper">
          <Icon icon="Releases" data={DataAsMarkup} />
        </div>
      </>
    )
    const tree = renderer.create(<IconWithData />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('render Icons with data and shadow', () => {
    const dataAsText = 'data as string'
    const DataAsMarkup = <>data as markup</>
    const IconWithShadowAndData = () => (
      <>
        <div className="">
          <Icon icon="SettingsOutline" hover={true} data={dataAsText} shadow='medium' />
        </div>
        <div className="">
          <Icon icon="SettingsOutline" hover={true} data={DataAsMarkup} shadow='medium' />
        </div>
      </>
    )
    const tree = renderer.create(<IconWithShadowAndData />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  // it('render Icons with active state prop', () => {
  //   const IconWithActive = () => (
  //     <>
  //       <div className="">
  //         <Icon icon='Preferences' active={true} hover={true} size="large" />
  //       </div>
  //     </>
  //   )
  //   const tree = renderer.create(<IconWithActive />).toJSON()
  //   expect(tree).toMatchSnapshot()
  // })
})
