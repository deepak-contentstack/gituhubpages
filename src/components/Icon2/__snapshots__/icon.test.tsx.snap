// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Icon render Icons with data 1`] = `
Array [
  <div
    className="flex IconStoriesWrapper"
  >
    <div
      className="flex-v-center"
      data-test-id="cs-Icon"
    >
      data as string
    </div>
  </div>,
  <div
    className="flex IconStoriesWrapper"
  >
    <div
      className="flex-v-center"
      data-test-id="cs-Icon"
    >
      data as markup
    </div>
  </div>,
]
`;

exports[`Icon render Icons with data and shadow 1`] = `
Array [
  <div
    className=""
  >
    <div
      className="flex-v-center"
      data-test-id="cs-Icon"
    >
      <div
        className="flex-center flex-v-center Icon__shadow-medium"
      />
      data as string
    </div>
  </div>,
  <div
    className=""
  >
    <div
      className="flex-v-center"
      data-test-id="cs-Icon"
    >
      <div
        className="flex-center flex-v-center Icon__shadow-medium"
      />
      data as markup
    </div>
  </div>,
]
`;

exports[`Icon render Icons with different icon state 1`] = `
Array [
  <div>
    ICONS DEFAULT :
  </div>,
  <div
    style={
      Object {
        "columns": "4",
        "width": "800px",
      }
    }
  >
    <div
      className="icon-spacing"
    />
    <div
      className="icon-spacing"
    />
    <div
      className="icon-spacing"
    />
    <div
      className="icon-spacing"
    />
    <div
      className="icon-spacing"
    />
    <div
      className="icon-spacing"
    />
  </div>,
  <div>
    ICONS HOVER :
  </div>,
  <div
    style={
      Object {
        "columns": "4",
        "width": "800px",
      }
    }
  >
    <div
      className="icon-spacing"
    />
    <div
      className="icon-spacing"
    />
    <div
      className="icon-spacing"
    />
    <div
      className="icon-spacing"
    />
    <div
      className="icon-spacing"
    />
    <div
      className="icon-spacing"
    />
  </div>,
  <div>
    ICONS PRESSED/SELECTED :
  </div>,
  <div
    style={
      Object {
        "columns": "4",
        "width": "800px",
      }
    }
  >
    <div
      className="icon-spacing"
    />
    <div
      className="icon-spacing"
    />
    <div
      className="icon-spacing"
    />
    <div
      className="icon-spacing"
    />
    <div
      className="icon-spacing"
    />
    <div
      className="icon-spacing"
    />
  </div>,
]
`;

exports[`Icon render default icons 1`] = `
<div
  style={
    Object {
      "columns": "4",
      "width": "800px",
    }
  }
>
  <div
    className="icon-spacing"
  />
  <div
    className="icon-spacing"
  />
  <div
    className="icon-spacing"
  />
  <div
    className="icon-spacing"
  />
  <div
    className="icon-spacing"
  />
  <div
    className="icon-spacing"
  />
</div>
`;

exports[`Icon render icon with fill property 1`] = `
<div>
  <div>
    Default Icon
  </div>
  <div>
    Icon with fill property
  </div>
</div>
`;

exports[`Icon render icon with height and width property 1`] = `
<div>
  <div>
    Default Icon
  </div>
  <div>
    Icon with height and width property
  </div>
</div>
`;

exports[`Icon render icon with stroke property 1`] = `
<div>
  <div>
    Default Icon
  </div>
  <div>
    Icon with stroke property
  </div>
</div>
`;

exports[`Icon render icons with different sizes 1`] = `
<div>
  <div>
     Tiny
  </div>
  <div>
     Small
  </div>
  <div>
     Large
  </div>
</div>
`;
